### PR TITLE
framework: Add doxygen figure to explain cache trackers

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
I couldn't really review #10979 until the tracker graph was documented in a consolidated visual figure.  This PR adds such a figure.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/11056)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
image

**Comment by jwnimmer-tri** [comment]
+@sherm1 for feature review, please.

I see now that the `#9171` labels are laid-out poorly.  I'll see if I can tune that up in the next push.

**Comment by SeanCurtis-TRI** [comment]
I had a similar concern about the lay out that you did. I fit can't be tuned into a better configuration, then I worry that it doesn't do its job very well.

I'd suggest a possible alternative of an authored SVG file. It's still ascii. We can directly control the visualization of the table in an effective manner. And it's probably at least as maintainable as the .dot file.

**Comment by jwnimmer-tri** [comment]
I was extremely happy with the layout, actually.  If you can try to explain what you found difficult, I can work on addressing it.

**Comment by SeanCurtis-TRI** [comment]
From my perspective:

1. There's no legend in the figure describing what the various lines look like (okay, that's a ticky-tacky detail).
2. We have more line crossing than necessary (`leaf_xcdot`) is a prime example of this phenomenon.
3. Currently, we distinguish between `diagram` and `leaf` quantities with label augementation. This would be better done as a visual cue (i.e., color). Then what stands out is the *quantity* and not the diagram/leaf dichotomy.
4. The black arrows only exist within one level (diagram or leaf). That means this graph is essentially two connected components with blue and red connections between them. Furthermore, the two sub-graphs are essentially identical. As such the parallels could be made much more apparent (perhaps via mirroring) with short, non-black-link-traversing red and blue connections.

Cleaning up all of these would make the graph-representation of the table far more effective than the posted version.

**Comment by jwnimmer-tri** [comment]
> There's no legend in the figure describing what the various lines look like (okay, that's a ticky-tacky detail).

That's in the Doxygen file instead of the dot file.  I should have made my screenshot slightly larger.

**Comment by jwnimmer-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/11056#-Lb0RN561_QVaMixaW4O:-Lb0RN571JvrFZH23KDS:btq1a2c) (no related file):*
@sherm1 There is something I'm not understanding.

For the energy/power trackers (`pe`, `ke`, `pc`, `pnc`), for all Systems they depend on the relevant sources for that System.  Per the TODOs in system.h, those dependencies are forecast to just be `configuration` / `kinematics` / etc (or still `all_sources` for `pnc`), but for now they depend on `all_sources`.  That's fine.

For the Diagram, in the cache_doxygen.h we _also_ claim to tell the diagram trackers for these guys to subscribe to all subsystems' trackers, etc:
```
pe                   | —                           | subsystem pe
ke                   | —                           | subsystem ke
pc                   | —                           | subsystem pc
pnc                  | —                           | subsystem pnc
```
If the diagram `pe` tracker already subscribes to the diagram `configuration` tracker, why does it _also_ need to subscribe to subsystems' `pe` trackers?  (Same for the rest of the energy / power.)  What could invalidate the `subsystem_5.pe` tracker but not the `diagram.configuration` tracker?

The same kind of questions applies to the diagram.xcdot tracker -- why does it subscribe to `subsystems.xcdot` trackers?

Basically I think `SubscribeDiagramCompositeTrackersToChildrens` list of `composites` commented as `// Cache entries.` don't make sense, since the well-known cache entries already always have their well-known Diagram dependencies.

---

**Comment by jwnimmer-tri** [comment]
I am reworking the diagram a bit based on those comments.  It's fair game to review the image and make more suggestions or defects, but FYI probably not worth looking at the `*.dot` contents yet in case I rejigger them.

**Comment by sherm1** [comment]
> Basically I think SubscribeDiagramCompositeTrackersToChildrens list of composites commented as // Cache entries. don't make sense, since the well-known cache entries already always have their well-known Diagram dependencies.

Issue #9205 (caching loose ends) has this entry:
- [ ] DiagramContext cache entries correctly subscribe to their constituent child subcontext cache entries, but also subscribe to the diagram-level sources (because the cache entries are set up by default like that). Those default dependencies should be deleted when adding in the subcontext dependencies.

I think this is what you're seeing there. There is some redundant wiring there, but the right fix is to leave the direct dependencies on the cache entries rather than on the prereqs. The reason is that a cache entry can serve as a "source" and can be modified directly even if its prereqs don't change. In that case I would like for the diagram composites to be invalidated, but they won't be if they are only monitoring cache entries indirectly through their prereqs.

Having redundant dependencies is a very minor issue (they'll get short-circuited) compared to _missing_ a dependency so I have been erring on the side of too many connections. Trimming down to a minimal set is an optimization I'd like to perform eventually.

**Comment by jwnimmer-tri** [comment]
Can you repost that in the correct discussion thread, please?

**Comment by jwnimmer-tri** [comment]
image

**Comment by SeanCurtis-TRI** [comment]
I like the updated diagram *a lot*. This morning, I posted a hand-crafted one from last night. I have no idea why it's not in this PR anymore... This has *not* been a good day for me and my computer.

**Comment by jwnimmer-tri** [comment]
Fixed (preview below).  Whether and how to show what comes "above" the Diagram trackers is a bit tricky.  I've added the input port subscription (arrow pointing up) to the superdiagram, but I haven't added the red arrows up to the superdiagram, and I haven't any added any arrows down from the superdiagram.  I think this is OK.

image

**Comment by jwnimmer-tri** [comment]
+@EricCousineau-tri for platform review, please.

**Comment by jwnimmer-tri** [comment]
Punting on debug and lsan; forcing in now.
